internal/controller/apiextensions/composite/connection.go

```diff
@@ -186,3 +168,188 @@ func (c *SecretStoreConnectionDetailsConfigurator) Configure(ctx context.Context
+type ConnectionDetailType string
+
+// ConnectionDetailType types.
+const (
+	ConnectionDetailTypeFromConnectionSecretKey ConnectionDetailType = "FromConnectionSecretKey"
+	ConnectionDetailTypeFromFieldPath           ConnectionDetailType = "FromFieldPath"
+	ConnectionDetailTypeFromValue               ConnectionDetailType = "FromValue"
+)
+
+// A ConnectionDetailExtractConfig configures how an XR connection detail should
+// be extracted.
+type ConnectionDetailExtractConfig struct {
+	// Type sets the connection detail fetching behaviour to be used. Each
+	// connection detail type may require its own fields to be set on the
+	// ConnectionDetail object.
+	Type ConnectionDetailType
+
+	// Name of the connection secret key that will be propagated to the
+	// connection secret of the composition instance.
+	Name string
+
+	// FromConnectionDetailKey is the key that will be used to fetch the value
+	// from the given target resource's connection details.
+	FromConnectionDetailKey *string
+
+	// FromFieldPath is the path of the field on the composed resource whose
+	// value to be used as input. Name must be specified if the type is
+	// FromFieldPath is specified.
+	FromFieldPath *string
+
+	// Value that will be propagated to the connection secret of the composition
+	// instance. Typically you should use FromConnectionSecretKey instead, but
+	// an explicit value may be set to inject a fixed, non-sensitive connection
+	// secret values, for example a well-known port.
+	Value *string
+}
+
+// ExtractConfigsFromTemplate builds extract configs for the supplied P&T style
+// composed resource template.
+func ExtractConfigsFromTemplate(t *v1.ComposedTemplate) []ConnectionDetailExtractConfig {
+	if t == nil {
+		return nil
+	}
+	out := make([]ConnectionDetailExtractConfig, len(t.ConnectionDetails))
+	for i := range t.ConnectionDetails {
+		out[i] = ConnectionDetailExtractConfig{
+			Type:                    connectionDetailType(t.ConnectionDetails[i]),
+			Value:                   t.ConnectionDetails[i].Value,
+			FromConnectionDetailKey: t.ConnectionDetails[i].FromConnectionSecretKey,
+			FromFieldPath:           t.ConnectionDetails[i].FromFieldPath,
+		}
+
+		if t.ConnectionDetails[i].Name != nil {
+			out[i].Name = *t.ConnectionDetails[i].Name
+		}
+
+		if out[i].Type == ConnectionDetailTypeFromConnectionSecretKey && out[i].FromConnectionDetailKey != nil {
+			out[i].Name = *out[i].FromConnectionDetailKey
+		}
+	}
+	return out
+}
+
+// ExtractConfigsFromDesired builds extract configs for the supplied Composition
+// Function desired state.
+func ExtractConfigsFromDesired(dr *iov1alpha1.DesiredResource) []ConnectionDetailExtractConfig {
+	if dr == nil {
+		return nil
+	}
+	out := make([]ConnectionDetailExtractConfig, len(dr.ConnectionDetails))
+	for i := range dr.ConnectionDetails {
+		out[i] = ConnectionDetailExtractConfig{
+			Type:                    ConnectionDetailType(dr.ConnectionDetails[i].Type),
```

Comment:
There is an inconsistency between `composite.ConnectionDetailType` and `iov1alpha1.ConnectionDetailTypeFromConnectionDetailKey`, i.e. `FromConnectionSecretKey` vs `FromConnectionDetailKey` which brakes extracting keys from connection details of managed resource in function pipelines. 

```suggestion
			Type:                    toConnectionDetailType(dr.ConnectionDetails[i].Type),
```

where

```
func toConnectionDetailType(detailType iov1alpha1.ConnectionDetailType) ConnectionDetailType {
	switch detailType {
	case iov1alpha1.ConnectionDetailTypeFromConnectionDetailKey:
		return ConnectionDetailTypeFromConnectionSecretKey
	default:
		return ConnectionDetailType(detailType)
	}
}
```

To continue with my experiments, I applied the above patch which may or may not what we want to do to fix the problem here. I would prefer being consistent with existing semantics but can also understand if we want to use this as an opportunity to use a better naming.

Reply:
> I would prefer being consistent with the existing API

On reflection I'm inclined to agree. Pushed add67cbb to address this.